File: pine-go/internal/runtime/parallel.go
package runtime

import (
	"context"
	"runtime/debug"
	"sync"

	"github.com/Liam0205/pineapple/pine-go/internal/types"
)

// splitInput divides an OperatorInput's items into n roughly equal shards,
// sharing the same common map. Returns the shard inputs and their start offsets
// within the original items slice.
func splitInput(input *types.OperatorInput, n int) ([]*types.OperatorInput, []int) {
	total := input.ItemCount()
	common := input.RawCommon()
	templated := input.RawTemplated()

	if n <= 1 || total == 0 {
		return []*types.OperatorInput{input}, []int{0}
	}
	if n > total {
		n = total
	}

	base := total / n
	rem := total % n

	parts := make([]*types.OperatorInput, n)
	offsets := make([]int, n)
	start := 0

	if input.IsLazy() {
		baseOffset := input.LazyOffset()
		frame := input.LazyFrame()
		itemDefaults := input.LazyItemDefaults()
		itemFields := input.LazyItemFields()
		for i := 0; i < n; i++ {
			size := base
			if i < rem {
				size++
			}
			parts[i] = types.NewLazyOperatorInput(common, frame, itemDefaults, itemFields, baseOffset+start, size)
			parts[i].SetTemplatedParams(templated)
			offsets[i] = start
			start += size
		}
	} else {
		items := input.RawItems()
		for i := 0; i < n; i++ {
			size := base
			if i < rem {
				size++
			}
			end := start + size
			shardItems := make([]map[string]any, size)
			copy(shardItems, items[start:end])
			parts[i] = types.NewOperatorInput(common, shardItems)
			parts[i].SetTemplatedParams(templated)
			offsets[i] = start
			start = end
		}
	}
	return parts, offsets
}

// mergeOutputs combines shard outputs into a single OperatorOutput by remapping
// item write indices using the provided offsets. Parallel ops are restricted
// to Transform without common_output (enforced by config validation), so
// shards may only emit item_writes, removed_items, and warnings. If any
// shard emits added_items, item_order, or common writes, we panic rather
// than silently drop — that path used to be a silent corruption surface.
func mergeOutputs(opName string, outputs []*types.OperatorOutput, offsets []int) *types.OperatorOutput {
	merged := types.NewOperatorOutput()

	for i, out := range outputs {
		if out == nil {
			continue
		}
		if len(out.GetAddedItems()) > 0 || out.GetItemOrder() != nil ||
			len(out.GetCommonWrites()) > 0 {
			panic(&types.PanicError{
				Operator: opName,
				Value:    "data_parallel shard emitted added_items, item_order, or common writes; only item_writes / removed_items / warnings are allowed",
			})
		}
		if w := out.GetWarning(); w != nil {
			merged.SetWarning(w)
		}
		offset := offsets[i]
		for _, w := range out.GetItemWrites() {
			merged.SetItem(w.Index+offset, w.Field, w.Value)
		}
		for localIdx := range out.GetRemovedItems() {
			merged.RemoveItem(localIdx + offset)
		}
	}
	return merged
}

// parallelExecute splits the input items into DataParallel shards, executes
// the operator concurrently on each shard, and merges the outputs.
func parallelExecute(ctx context.Context, cop *CompiledOperator, input *types.OperatorInput) (*types.OperatorOutput, error) {
	parts, offsets := splitInput(input, cop.Config.DataParallel)

	if len(parts) == 1 {
		out := types.NewOperatorOutput()
		err := executeWithRecovery(ctx, cop, parts[0], out)
		return out, err
	}

	execCtx, cancel := context.WithCancel(ctx)
	defer cancel()

	outputs := make([]*types.OperatorOutput, len(parts))
	var (
		wg      sync.WaitGroup
		errOnce sync.Once
		first   error
	)

	for i := range parts {
		wg.Add(1)
		go func(idx int) {
			defer wg.Done()
			if execCtx.Err() != nil {
				return
			}
			out := types.NewOperatorOutput()
			err := executeWithRecovery(execCtx, cop, parts[idx], out)
			if err != nil {
				errOnce.Do(func() {
					first = err
					cancel()
				})
				return
			}
			outputs[idx] = out
		}(i)
	}

	wg.Wait()
	if first != nil {
		return nil, first
	}
	return mergeOutputs(cop.Name, outputs, offsets), nil
}

// executeWithRecovery runs a single operator Execute with panic recovery.
func executeWithRecovery(ctx context.Context, cop *CompiledOperator, input *types.OperatorInput, output *types.OperatorOutput) (err error) {
	defer func() {
		if r := recover(); r != nil {
			err = &types.PanicError{
				Operator: cop.Name,
				Value:    r,
				Stack:    string(debug.Stack()),
			}
		}
	}()
	return cop.Instance.Execute(ctx, input, output)
}

File: pine-python/pine/parallel.py
"""Parallel (sharded) operator execution logic."""
from __future__ import annotations

import threading
from typing import Any

from pine.cancellation import CancellationToken
from pine.errors import PanicError
from pine.operator import Operator, OperatorInput, OperatorOutput


def _parallel_execute(
    token: CancellationToken,
    operator: Operator,
    input_: OperatorInput,
    data_parallel: int,
    op_name: str,
) -> OperatorOutput:
    """Execute an operator in parallel shards over items.

    Splits items into N shards, executes each with a child CancellationToken,
    and merges outputs. Only called for ConcurrentSafe operators.
    """
    items = input_.raw_items()
    item_count = len(items)
    shard_count = min(data_parallel, max(item_count, 1))

    if shard_count <= 1:
        output = OperatorOutput()
        operator.execute(token, input_, output)
        return output

    # Split items into shards
    shards: list[list[dict[str, Any]]] = [[] for _ in range(shard_count)]
    shard_indices: list[list[int]] = [[] for _ in range(shard_count)]
    for i, item in enumerate(items):
        shard_idx = i % shard_count
        shards[shard_idx].append(item)
        shard_indices[shard_idx].append(i)

    # Execute each shard with a child token
    shard_outputs: list[OperatorOutput | None] = [None] * shard_count
    shard_errors: list[Exception | None] = [None] * shard_count
    child_token = token.child()

    def _execute_shard(shard_i: int):
        if token.is_cancelled() or child_token.is_cancelled():
            return
        shard_input = OperatorInput(input_.raw_common(), shards[shard_i])
        shard_out = OperatorOutput()
        try:
            operator.execute(child_token, shard_input, shard_out)
            shard_outputs[shard_i] = shard_out
        except Exception as e:
            shard_errors[shard_i] = e
            child_token.cancel()

    threads = []
    for si in range(shard_count):
        t = threading.Thread(target=_execute_shard, args=(si,), daemon=True)
        threads.append(t)
        t.start()
    for t in threads:
        t.join()

    # Check for shard errors
    for err in shard_errors:
        if err is not None:
            raise err

    # Merge shard outputs back into a single OperatorOutput
    merged = OperatorOutput()
    for si in range(shard_count):
        shard_out = shard_outputs[si]
        if shard_out is None:
            continue

        # Parallel ops are restricted to Transform without common_output
        # (enforced by config validation); shards may only emit item_writes,
        # removed_items, and warnings. Anything else is a silent-drop
        # corruption surface — raise loudly instead.
        if (shard_out.added_items or shard_out.item_order is not None or
                shard_out.common_writes):
            raise PanicError(
                f'pine: panic in operator "{op_name}": '
                f'data_parallel shard emitted added_items, item_order, '
                f'or common writes; only item_writes / removed_items / '
                f'warnings are allowed'
            )

        # Merge item writes (remap shard-local indices to global)
        for local_idx, writes in shard_out.item_writes.items():
            if local_idx < len(shard_indices[si]):
                global_idx = shard_indices[si][local_idx]
                for field_name, value in writes.items():
                    merged.set_item(global_idx, field_name, value)

        # Merge removed items
        for local_idx in shard_out.removed_items:
            if local_idx < len(shard_indices[si]):
                global_idx = shard_indices[si][local_idx]
                merged.remove_item(global_idx)

        # Merge warning
        if shard_out.warning is not None:
            merged.set_warning(shard_out.warning)

    return merged
